compiler/rustc_session/src/session.rs

```diff
@@ -1723,59 +1723,123 @@ pub enum IncrCompSession {
-            json_rendered,
-            None,
-            fallback_bundle,
-            None,
-            false,
-            false,
-            TerminalUrl::No,
-        )),
-    };
-    rustc_errors::Handler::with_emitter(true, None, emitter)
+/// A wrapper around an [`Handler`] that is used for early error emissions.
+pub struct EarlyErrorHandler {
+    handler: Handler,
 }
 
-#[allow(rustc::untranslatable_diagnostic)]
-#[allow(rustc::diagnostic_outside_of_impl)]
-#[must_use = "ErrorGuaranteed must be returned from `run_compiler` in order to exit with a non-zero status code"]
-pub fn early_error_no_abort(
-    output: config::ErrorOutputType,
-    msg: impl Into<DiagnosticMessage>,
-) -> ErrorGuaranteed {
-    early_error_handler(output).struct_err(msg).emit()
-}
+impl EarlyErrorHandler {
+    pub fn new(output: ErrorOutputType) -> Self {
+        // FIXME(#100717): early errors aren't translated at the moment, so this is fine, but it will
+        // need to reference every crate that might emit an early error for translation to work.
+        let fallback_bundle =
+            fallback_fluent_bundle(vec![rustc_errors::DEFAULT_LOCALE_RESOURCE], false);
+        let emitter: Box<dyn Emitter + sync::Send> = match output {
+            config::ErrorOutputType::HumanReadable(kind) => {
+                let (short, color_config) = kind.unzip();
+                Box::new(EmitterWriter::stderr(
+                    color_config,
+                    None,
+                    None,
+                    fallback_bundle,
+                    short,
+                    false,
+                    None,
+                    false,
+                    false,
+                    TerminalUrl::No,
+                ))
+            }
+            config::ErrorOutputType::Json { pretty, json_rendered } => {
+                Box::new(JsonEmitter::basic(
+                    pretty,
+                    json_rendered,
+                    None,
+                    fallback_bundle,
+                    None,
+                    false,
+                    false,
+                    TerminalUrl::No,
+                ))
+            }
+        };
 
-#[allow(rustc::untranslatable_diagnostic)]
-#[allow(rustc::diagnostic_outside_of_impl)]
-pub fn early_error(output: config::ErrorOutputType, msg: impl Into<DiagnosticMessage>) -> ! {
-    early_error_handler(output).struct_fatal(msg).emit()
-}
+        Self { handler: rustc_errors::Handler::with_emitter(true, None, emitter) }
+    }
+
+    pub fn abort_if_errors(&self) {
+        self.handler.abort_if_errors()
+    }
+
+    /// Swap out the underlying handler once we acquire the user's preference on error emission
+    /// format. Any errors prior to that will cause an abort and all stashed diagnostics of the
+    /// previous handler will be emitted.
+    pub fn abort_if_error_and_set_error_format(&mut self, output: ErrorOutputType) {
+        self.handler.abort_if_errors();
+
+        let fallback_bundle =
```

Comment:
Is this part duplicated with `new`? If it is, could you extract the emitter creation into a function?